### DISCUSSION TITLE
What's your workflow for maintaining the patches?

### DISCUSSION THREAD

**Discussion by mpawlowski-eyeo** [bromite/bromite#2184] [General]
As Chromium gets updated, the patches sometimes need to be modified to keep working. I'd assume there are some breakages after most major version updates.

How do you go about maintaining the patches? When you prepare patches for a newer version of Chromium, do you apply all of them in order, see what needs fixing and then *manually* select which changed lines should go to which patch file? Do you apply patches one by one, fix code and build after every patch, run the browser/tests to see if the patch works?

How do you avoid attributing changes to wrong patch files, how do you handle refactorings that affect multiple patches? Is there a set of tools to help with your way of work?

**Comment by uazo** [comment]
I have much less experience of @csagan5, I do:
1. launch my [tool](https://github.com/uazo/bromite/blob/ci/.github/workflows/check-git-apply.yaml) that apply all patches on a new version, or I use [apply-all-patch.sh](https://github.com/uazo/bromite-buildtools/blob/master/apply-all-patch.sh) manually which also uses wiggle if the apply is not automatic
2. if even wiggle can't do it, two possibilities:
2.1 use [superpatch](https://github.com/uazo/superpatch) to understand what the patch was in the old code and in case, and, if the change is not impactful, I fix it directly
2.2 comment the code by inserting a compile error, and think about it later

the environment I use is visual studio code with gitlens, which I find very useful, because is able to understand the results of wiggle, allowing me to apply changes with simple mouse clicks.

once all the patches have been applied, I launch the build and unfortunately each error must be fixed by hand, however, fortunately, there are never many, and more or less, it's the usual patches that need to be fixed.
then I assign every change to the patches through fixups and I rebase with the export of the final patches.

once finished, I manually check each patch one by one with the previous version, looking for lost changes, and if there are, I start over. I know I can automate this thing, but I haven't done it yet (it is tedious for me to make these tools, I have more fun looking for new things in the chromium code :).
for the small patches (or that I know or I've written) I do some tests for the new changes (always manually, sigh). here too, I would like to automate the tests also because unfortunately I don't do tests for the bigger ones (I hope to be able to simplify them sooner or later). here too, I have already seen a way to do it but it is long and complex.

at the end, I launch the latest final build with the generation of the list of new flags and [check them](https://github.com/uazo/bromite-flags-list/blob/main/flags-to-check) one by one. for v103 I haven't done it yet because keeping a branch not aligned with the work of csagan5 has always turned out to be a double / triple job, so I learned to wait.

> and build after every patch,

I tried, it's not worth it

> How do you avoid attributing changes to wrong patch files

manually, but gitlens helps me figure it out

> how do you handle refactorings that affect multiple patches? 

again, manually

> Is there a set of tools to help with your way of work?

apart from the ones I created by myself, [code search](https://source.chromium.org/) and [csagan5 mtool](https://github.com/bromite/mtool) it is very useful and I use it very often

**Comment by mpawlowski-eyeo** [comment]
Thanks for the detailed reply :) Do you feel like working with multiple patches (that all need to be applied anyway, and in the right order) is better than just having a one big "Bromite patch"?

**Reply by uazo** [reply]
> is better than just having a one big "Bromite patch"?

absolutely better.
in case of a bug it is easier to understand which changes have been made.

allow me, now I ask a question: why are you so interested?

**Reply by mpawlowski-eyeo** [reply]
We're maintaining a fork of Chromium too. I'm researching ways to make changes in our fork easier to consume for third party developers, and Bromite's patch files definitely seem easier to apply to an existing codebase. But it seems harder to maintain, with all the manual upkeep of individual patches, than a git branch.

I'm trying to assess if that's indeed the case, a trade-off between ease of maintenance (on the "producer" side) vs ease of integration (on the "consumer" side).

**Reply by uazo** [reply]
interesting, keep us updated and have a nice work!

> We're maintaining a fork of Chromium too

Something that I can peek at?

**Reply by mpawlowski-eyeo** [reply]
Sure: https://gitlab.com/eyeo/adblockplus/abpchromium

It adds ad-blocking that's not based on Chromium's subresource filtering and has feature parity with the Adblock Plus extension.

**Reply by uazo** [reply]
> It adds ad-blocking that's not based on Chromium's subresource filtering and has feature parity with the Adblock Plus extension.

sure, I think it needs no introduction! :)
I apologize if I seemed to know-it-all, I was hoping for a new collaborator... :)

> Sure: https://gitlab.com/eyeo/adblockplus/abpchromium

It is the same library that I saw in the "sources" of edge (@csagan5 do you remember?), have you tried asking them?

>  I'm researching ways to make changes in our fork easier to consume for third party developers.

however at this point I rephrase my answer, because, compared to bromite, the purposes seem different to me, where we modify (heavily at times) chromium, often removing rather than adding.
when I developed the userscript I asked myself the same question you did and I thought the path of using both methods (branch + patch) was viable.
my idea was to create a gyp that applies the changes to the chromium codebase as a patch, while using the gsync hooks to fetch the repo: I didn't do this due to lack of resources and because it would only apply to userscripts in bromite.
but this would allow more developers to work normally with git (and above all to have a history of the changes, which the patches hardly expose due to the changes to the numbering of the lines) while consumers would just need to add a dep to their gn (theoretically).
is it a way that you have verified?

**Reply by mpawlowski-eyeo** [reply]
hey, forgot to reply, vacation ;)

The "script that applies patches during gsync" feels like something that:
- requires extra work to maintain, and can break in hilarious ways eg. when someone already has a dirty repo
- isn't transparent, the dev that wants to use our "SDK" doesn't see what will get applied
- can be tricky when you want to apply a diff or in all the weird use cases people have when working with the chromium codebase (rebasing, commiting half-done merges, auto-rebasing CI tools etc)
- reinvents git in some aspects?

The benefit is that it's one-click *if it works*.

but I didn't think of that before, it's one idea to add to our list of things to check

**Reply by uazo** [reply]
> hey, forgot to reply, vacation ;)

but imagine!

> feels like something that:

yes, you are probably right, actually other forks don't just use the patch system, but for us it might be feasible, because starting from chromium or starting from abpchromium would be the same.
it is definitely something that I will try, too bad to have so many things to do/try and little time.

**Reply by mpawlowski-eyeo** [reply]
> > Sure: https://gitlab.com/eyeo/adblockplus/abpchromium
> > It adds ad-blocking that's not based on Chromium's subresource filtering and has feature parity with the Adblock Plus extension.
> 
> nice to see i want to test it can you give us the way to download your app or need to compiled myself.

You need to compile it yourself, our build-making CI is not made public. The build instructions are the same as for normal Chromium and afair we don't need special gn args or extra deps. It builds on Android, Linux, MacOs and Windows.

**Reply by uazo** [reply]
> You need to compile it yourself,

tried! actually it was easy to get adblock-plus (without telemetry :) into bromite.

![image](https://user-images.githubusercontent.com/29201891/188208890-601bb58c-4ab8-400f-8446-b449f8ef8cd0.png)

the one on the left is also bromite!

**Reply by csagan5** [reply]
You mean the extension?

**Reply by uazo** [reply]
no, the code extracted from https://gitlab.com/eyeo/adblockplus/chromium-sdk transformed into a patch and applied on bromite.
are we interested in including it in bromite? meanwhile, is it legally possible?

**Reply by mpawlowski-eyeo** [reply]
As long as you're ok with the GPL license, I think it's legally OK from our POV :)

FWIW, this SDK typically performs much better than either the extension or Chromium's built-in subresource filter mechanism.

**Reply by uazo** [reply]
well then, you tell me @csagan5.
it can also be just a temporary test. to set up the thing is not very complex, the steps I see:
1) generate the "base" patch automatically with an ad hoc repo and actions
which consists of the checkout of eyeo master and relative chromium branch, copy the .git folder and extract the patch.
2) create another patch that disables telemetry and downloading rules from their servers (as well as a few pings here and there)
perhaps to a more accurate control than also other to be reviewed, I have not yet seen well
3) check SubscriptionCollectionImpl::GetRewriteUrl() because it redirects the requests according to the filter rule, which probably needs to be checked

**Reply by csagan5** [reply]
License is here: https://gitlab.com/eyeo/adblockplus/chromium-sdk/-/blob/eyeo-107-dev/components/adblock/LICENSE (GPLv3), so it is not a problem.

> to set up the thing is not very complex, the steps I see:

You can start using that and then submit the patches, so that we always have a review by humans involved. I assume this component does not use the subresource filter at all, so we would need to remove that from Bromite.

**Reply by csagan5** [reply]
Related: https://github.com/bromite/bromite/issues/2319

This video blocker should be also open source.

**Reply by uazo** [reply]
> You can start using that and then submit the patches

yes, I'm already carrying on, I'll handle that.
here the [repo](https://github.com/uazo/eyeo-adblockplus-as-a-patch) that automatically creates the patch from their sources, but I don't understand which branch to use, @mpawlowski-eyeo any suggestions? how to understand which is the stable?

> I assume this component does not use the subresource filter at all, so we would need to remove that from Bromite.

do you mean the patch that updates the filters in bromite? I tried and they don't bother each other, we can think about it later.

> This video blocker should be also open source.

their adblock as it is already blocks them

**Reply by mpawlowski-eyeo** [reply]
> yes, I'm already carrying on, I'll handle that. here the [repo](https://github.com/uazo/eyeo-adblockplus-as-a-patch) that automatically creates the patch from their sources, but I don't understand which branch to use, @mpawlowski-eyeo any suggestions? how to understand which is the stable?

Normally, the way we work is:
- The default branch on gitlab is our current development branch. As of writing, its `eyeo-107-dev`. It's not stable.
- We switch to a newer major Chromium version at [Branch point](https://chromium.googlesource.com/chromium/src.git/+/main/docs/process/release_cycle.md#branch-point). If you look at https://chromiumdash.appspot.com/schedule it says for 108 `Branch Thu, Oct 13, 2022`.
- On that day, we'll create `eyeo-108-dev` by rebasing changes from `eyeo-107-dev` on top of current upstream 108 dev. If there are major conflicts to fix, they'll be fixed here. We'll then make `eyeo-108-dev` our default development branch. It's also not stable.
- At the same time (Oct 13) we'll create `eyeo-107-beta`, by rebasing changes from `eyeo-107-dev` over newest usptream 107 beta. Devs will fix any conflicts (usually none or few), and QAs will start manual testing.
- Once testing is done, we will add a git tag to the final commit of `eyeo-107-beta` branch, `eyeo-beta-107.0.xxx-Y.Z`. See some tags here: https://gitlab.com/eyeo/adblockplus/chromium-sdk/-/tags 
- *Monitoring tags is one way to automate getting notified about releases*.
- Usually around 2 weeks later the upstream 107 branch will have its Stable release. The schedule says, for 107, `Stable Release Tue, Oct 25, 2022`. 
- At that time, we will rebase our changes from `eyeo-107-beta` into `eyeo-107-rc` on top of the current stable upstream 107 and do some final QA.
- Once we finish testing the stable build, we'll again make a tag: `eyeo-rc-107.yyyy-Y.Z` for the final commit on `eyeo-107-rc`.
- We usually won't update the `beta` or `rc` branch after the final commit receives the tag, unless we find a critical error. In that case, we create a new tag.

We only do the stable ("RC") part of this scenario for *odd* versions, so 105, 107, 109 etc. Even versions (108, 110, 112) stop at the *beta tag*, and we only run automated tests on them, no manual QA. We do this to preserve QA resources. So there will be an `eyeo-108-dev`, then `eyeo-108-beta`, the beta branch will finalize with something like an `eyeo-beta-108.0.111.222-5.0` tag, and there will NOT be an `eyeo-108-rc`.

Another way of learning about releases is that we send out emails with release announcements to our partners - we can probably add your emails to our list. We usually send the emails immediately after tagging a beta or stable release.

We also publish [Release notes](https://gitlab.com/eyeo/adblockplus/chromium-sdk/-/wikis/Release-notes) when we have a new stable version. They don't get published for even versions, because those stop at beta.

Our code on `eyeo-xxx-beta` branch it's frozen, it only receives bugfixes cherry-picked from the development branch. For the purpose of using this code in your codebase, *you can consider our beta branches as stable branches*, the caveat is that they're based on *upstream* beta code. The *adblocking* code, or the patch you create, will likely be the same on `eyeo-107-beta` and `eyeo-107-rc`, when compared to their respective upstreams.

Let me know if that works for you or if you have improvement ideas

**Reply by uazo** [reply]
thank you very much for the directions! 

> Let me know if that works for you or if you have improvement ideas

I think I'll go step by step, no idea for now, but if it comes to me I'll bother you again :)
I keep you updated

**Reply by mpawlowski-eyeo** [reply]
BTW @uazo, @csagan5 would you be interested in attending https://adfilteringdevsummit.com/ ? Either remote or in-person. I'm gonna be there and would be happy to meet you :)

**Reply by uazo** [reply]
looks interesting, thanks for the tip, although I would love to take a trip to amsterdam, the costs are prohibitive for a hobby
about online/remote, I know myself and I would not understand anything due my bad english... maybe I begin to see those of 2021 on youtube waiting for the new ones
thanks anyway!

**Comment by csagan5** [comment]
> I'd assume there are some breakages after most major version updates.

For most patches not, but for some big patches there is breakage for every major version update.

> How do you go about maintaining the patches? When you prepare patches for a newer version of Chromium, do you apply all of them in order, see what needs fixing and then manually select which changed lines should go to which patch file? Do you apply patches one by one, fix code and build after every patch, run the browser/tests to see if the patch works?

I fix patches that do not apply manually and then at the end (once all patches are applied) I start to make fixups to fix building/linking/crash issues. Once that is complete I squash the fixups.

So builds are not made after every patch, that would be ideal but not maintainable for 1-2 persons.

> How do you avoid attributing changes to wrong patch files, how do you handle refactorings that affect multiple patches? Is there a set of tools to help with your way of work?

I check history of the individual files and attribute them second the purpose of the change/patch. I am developing some tooling that helps with this process but it is not yet released both because it is not ready and because I am afraid that it would create an endless stream of support requests for the builds without actually helping Bromite development (debatable). I have not decided yet about it, will continue developing it for the time being.
